trade-finance-manager-ui/server/controllers/case/cancellation/reason-for-cancelling.controller.get.test.ts

```diff
@@ -0,0 +1,114 @@
+import { createMocks } from 'node-mocks-http';
+import { DEAL_SUBMISSION_TYPE } from '@ukef/dtfs2-common';
+import { aTfmSessionUser } from '../../../../test-helpers';
+import { getReasonForCancelling, GetReasonForCancellingRequest } from './reason-for-cancelling.controller';
+import { PRIMARY_NAVIGATION_KEYS } from '../../../constants';
+import { ReasonForCancellingViewModel } from '../../../types/view-models';
+import api from '../../../api';
+import { canSubmissionTypeBeCancelled } from '../../helpers';
+
+jest.mock('../../../api', () => ({
+  getDeal: jest.fn(),
+  getDealCancellation: jest.fn(),
+}));
+
+jest.mock('../../helpers', () => ({
+  canSubmissionTypeBeCancelled: jest.fn(() => true),
```

Comment:
Do we need to mock this, can the actual function be used rather?

Reply:
Since this is a unit test I think it makes sense to mock this dependency & rely on the unit tests for `canSubmissionTypeBeCancelled` to check AIN/MIA.

Reply:
Thing is @MarRobSoftwire it's not representative of what it's actually doing. For example, with the current approach, if I change the mock for `api.getDeal` to return a deal with an MIA `submissionType`, the `it('renders the reason for cancelling page'` test will still pass, because `canSubmissionTypeBeCancelled` is marked to return true regardless of the `submissionType`. Outside of a unit test, the page would not render and it'd redirect instead

Reply:
Just weighing in some thoughts here -- not sure if any of this has already been said.

My understanding is that a unit test should test the smallest isolatable testable unit incredibly thoroughly. The `unit` in this case is a function.

The question over mocking external areas is always interesting. I always lean to mock everything that we're importing (exceptions being external modules). This is because:
1. We test these imported `units` elsewhere with 100% coverage (if we don't, as part of this PR we should).
2. The tests should not care about the implementation of other units, otherwise these stop being unit tests and start being half unit, half api tests. It should only care that `given I provide a function with x, it sends back y`
3. Testing these units here does not add any additional coverage to our testing as long as point 1 above is true (and it should be)

We should look to have a test double folder for these mocks to avoid having to repeat them.

My thoughts are in the screenshot below -- with the red showing what we should unit test.

![{D17CBDD5-D13F-4299-A975-54A6D9701800}](https://github.com/user-attachments/assets/0380f104-9ad5-43ae-b280-b61d9b71f164)

This would achieve 100% code coverage as this ethos of smallest possible testing unit is continued throughout the codebase

![{F71900A9-6505-4FF0-9A21-BC9989D184BC}](https://github.com/user-attachments/assets/dcbec95a-f459-47fa-969b-3c6a85151405)

I think not mocking helpers is a bit odd -- as we have unit tests which look like this, and covers more than a minimal tested unit: in purple 
![{22FA3D1E-1DAF-426A-9D7D-BD57272ADD52}](https://github.com/user-attachments/assets/40d604dc-6340-49f3-9a87-8cdef754ed7c)

Reply:
alternatively, if we are wanting to test this, which I think is a bit odd as effectively we have unit tests relying on the implementation of imported functions (and all the reasons above), we should 100% look to, as part of creating helpers, have a `withBlaHelper({makeRequest, addBlaParameterToRequest: (param)=> {baseReq, field: param}})` and then have all the tests in a helper to keep tests dry

Reply:
Thanks @AlexBramhill  for the above detailed explanation.
I can see the helper function is quite a simple implementation

```typescript
export const canSubmissionTypeBeCancelled = (submissionType: DealSubmissionType): boolean => {
  return submissionType === AIN || submissionType === MIN;
}
```

Ideally I would have suggested to mock a database call or an expensive external API call but given this function merely ascertain whether the deal submission type is either notice or not I think mocking is not required.

Though agree with unit test coverage but mocking a simple implementation will not harm based on mock deal it should return `true` anyways. This also works well for future refactorisation of this controller if the scope is expanded and we will also need to cater of deal type and its associated facility types.

Reply:
Sounds great to me

Reply:
@abhi-markan @ttbarnes I've refactored the tests to no longer mock `canSubmissionTypeBeCancelled` 👍

Reply:
Thank you both, will mark it as resolved.